File: defence_claw/defenseclaw/cli/defenseclaw/enforce/plugin_enforcer.py
# Copyright 2026 Cisco Systems, Inc. and its affiliates
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0

"""PluginEnforcer — filesystem quarantine for plugins.

Mirrors internal/enforce/plugin_enforcer.go.
"""

from __future__ import annotations

import os
import shutil

from defenseclaw.inventory.plugin_identity import (
    PluginIdentityError,
    canonical_plugin_id,
    filesystem_identity_key,
    is_link_or_reparse,
    validate_plugin_id,
)


class PluginEnforcer:
    def __init__(self, quarantine_dir: str) -> None:
        self.quarantine_dir = os.path.join(quarantine_dir, "plugins")
        os.makedirs(self.quarantine_dir, exist_ok=True)

    @staticmethod
    def _safe_segment(value: str) -> str | None:
        safe = os.path.basename(value)
        if not safe or safe != value or safe in (".", ".."):
            return None
        return safe

    def _quarantine_path(self, plugin_name: str, connector: str = "") -> str | None:
        safe_name = self._safe_segment(plugin_name)
        if safe_name is None:
            return None
        if connector:
            safe_connector = self._safe_segment(connector)
            if safe_connector is None:
                return None
            dest = os.path.join(self.quarantine_dir, safe_connector, safe_name)
        else:
            dest = os.path.join(self.quarantine_dir, safe_name)
        if not os.path.realpath(dest).startswith(os.path.realpath(self.quarantine_dir) + os.sep):
            return None
        return dest

    def quarantine(
        self,
        plugin_name: str,
        source_path: str,
        connector: str = "",
    ) -> str | None:
        """Move plugin directory to quarantine. Returns quarantine path or None."""
        if is_link_or_reparse(source_path):
            return None
        real_path = os.path.realpath(source_path)
        if not os.path.exists(real_path):
            return None
        try:
            source_id, _manifest = canonical_plugin_id(source_path)
            if filesystem_identity_key(source_id, os.path.dirname(source_path)) != filesystem_identity_key(
                validate_plugin_id(plugin_name), os.path.dirname(source_path)
            ):
                return None
        except PluginIdentityError:
            return None
        dest = self._quarantine_path(plugin_name, connector)
        if dest is None:
            return None
        os.makedirs(os.path.dirname(dest), exist_ok=True)
        if os.path.lexists(dest):
            return None
        shutil.move(real_path, dest)
        return dest

    def restore(
        self,
        plugin_name: str,
        restore_path: str,
        allowed_roots: list[str] | None = None,
        connector: str = "",
    ) -> bool:
        """Restore a quarantined plugin to its original location."""
        src = self._quarantine_path(plugin_name, connector)
        if src is None:
            return False
        if is_link_or_reparse(src) or not os.path.exists(src):
            return False
        try:
            source_id, _manifest = canonical_plugin_id(src)
            if filesystem_identity_key(source_id, os.path.dirname(src)) != filesystem_identity_key(
                validate_plugin_id(plugin_name), os.path.dirname(src)
            ):
                return False
        except PluginIdentityError:
            return False
        real_dest = os.path.realpath(restore_path)
        if allowed_roots:
            if not any(
                real_dest == os.path.realpath(r) or real_dest.startswith(os.path.realpath(r) + os.sep)
                for r in allowed_roots
            ):
                return False
        if os.path.lexists(restore_path):
            return False
        os.makedirs(os.path.dirname(restore_path), exist_ok=True)
        shutil.move(src, restore_path)
        return True

    def is_quarantined(self, plugin_name: str, connector: str = "") -> bool:
        path = self._quarantine_path(plugin_name, connector)
        return bool(path and os.path.exists(path))

File: defence_claw/defenseclaw/internal/enforce/plugin_enforcer.go
// Copyright 2026 Cisco Systems, Inc. and its affiliates
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0

package enforce

import (
	"fmt"
	"os"
	"path/filepath"

	"github.com/defenseclaw/defenseclaw/internal/sandbox"
)

type PluginEnforcer struct {
	quarantineDir string
	shell         *sandbox.OpenShell
}

func NewPluginEnforcer(quarantineDir string, shell *sandbox.OpenShell) *PluginEnforcer {
	return &PluginEnforcer{quarantineDir: quarantineDir, shell: shell}
}

func (e *PluginEnforcer) Quarantine(pluginPath string) (string, error) {
	// Refuse bundled-skill containers and descendants before any stat
	// touches the source tree — a vendor-supplied .system container
	// must never be moved to the quarantine store even in the
	// crash-recovery reconciler. The check is on the caller-supplied
	// path (not a resolved one) because plugin_enforcer operates on
	// paths the caller has already validated; the sibling
	// AssetQuarantinePlan path does its own containment check and
	// re-verifies here as defence in depth.
	if IsBundledSkillPath(pluginPath) {
		return "", ErrBundledSkill
	}
	info, err := os.Stat(pluginPath)
	if err != nil {
		return "", fmt.Errorf("enforce: plugin path %q: %w", pluginPath, err)
	}

	name := filepath.Base(pluginPath)
	dest := filepath.Join(e.quarantineDir, "plugins", name)

	if err := os.MkdirAll(filepath.Dir(dest), 0o700); err != nil {
		return "", fmt.Errorf("enforce: create quarantine dir: %w", err)
	}

	if info.IsDir() {
		if err := copyDir(pluginPath, dest); err != nil {
			return "", fmt.Errorf("enforce: copy plugin to quarantine: %w", err)
		}
		if err := os.RemoveAll(pluginPath); err != nil {
			return dest, fmt.Errorf("enforce: remove original plugin: %w", err)
		}
	} else {
		if err := copyFile(pluginPath, dest); err != nil {
			return "", fmt.Errorf("enforce: copy plugin to quarantine: %w", err)
		}
		if err := os.Remove(pluginPath); err != nil {
			return dest, fmt.Errorf("enforce: remove original plugin: %w", err)
		}
	}

	return dest, nil
}

func (e *PluginEnforcer) Restore(pluginName, originalPath string) error {
	base := filepath.Base(pluginName)
	src := filepath.Join(e.quarantineDir, "plugins", base)
	if _, err := os.Stat(src); err != nil {
		return fmt.Errorf("enforce: quarantined plugin %q not found: %w", base, err)
	}

	info, err := os.Stat(src)
	if err != nil {
		return fmt.Errorf("enforce: stat quarantined plugin: %w", err)
	}

	if info.IsDir() {
		if err := copyDir(src, originalPath); err != nil {
			return fmt.Errorf("enforce: restore plugin: %w", err)
		}
		return os.RemoveAll(src)
	}

	if err := copyFile(src, originalPath); err != nil {
		return fmt.Errorf("enforce: restore plugin: %w", err)
	}
	return os.Remove(src)
}

func (e *PluginEnforcer) IsQuarantined(pluginName string) bool {
	base := filepath.Base(pluginName)
	path := filepath.Join(e.quarantineDir, "plugins", base)
	_, err := os.Stat(path)
	return err == nil
}
